discovery/uyuni/uyuni.go

```diff
@@ -0,0 +1,319 @@
+		interval: time.Duration(conf.RefreshInterval),
+		sdConfig: conf,
+		logger:   logger,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"uyuni",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d
+}
+
+func (d *Discovery) getEndpointLabels(
+	endpoint endpointInfo,
+	systemGroupIDs []systemGroupID,
+	networkInfo networkInfo,
+) model.LabelSet {
+
+	var addr string
+	managedGroupNames := getSystemGroupNames(systemGroupIDs)
+	addr = fmt.Sprintf("%s:%d", networkInfo.Hostname, endpoint.Port)
+
+	result := model.LabelSet{
+		model.AddressLabel:       model.LabelValue(addr),
+		uyuniLabelMinionHostname: model.LabelValue(networkInfo.Hostname),
+		uyuniLabelPrimaryFQDN:    model.LabelValue(networkInfo.PrimaryFQDN),
+		uyuniLablelSystemID:      model.LabelValue(fmt.Sprintf("%d", endpoint.SystemID)),
+		uyuniLablelGroups:        model.LabelValue(strings.Join(managedGroupNames, d.sdConfig.Separator)),
+		uyuniLablelEndpointName:  model.LabelValue(endpoint.EndpointName),
+		uyuniLablelExporter:      model.LabelValue(endpoint.ExporterName),
+		uyuniLabelProxyModule:    model.LabelValue(endpoint.Module),
+		uyuniLabelMetricsPath:    model.LabelValue(endpoint.Path),
+	}
+
+	level.Debug(d.logger).Log("msg", "Configured target", "Labels", fmt.Sprintf("%+v", result))
+
+	return result
+}
+
+func getSystemGroupNames(systemGroupsIDs []systemGroupID) []string {
+	managedGroupNames := make([]string, 0, len(systemGroupsIDs))
+	for _, systemGroupInfo := range systemGroupsIDs {
+		managedGroupNames = append(managedGroupNames, systemGroupInfo.GroupName)
+	}
+
+	return managedGroupNames
+}
+
+func (d *Discovery) getTargetsForSystems(
+	rpcClient *xmlrpc.Client,
+	token string,
+	entitlement string,
+) ([]model.LabelSet, error) {
+
+	result := make([]model.LabelSet, 0)
+
+	systemGroupIDsBySystemID, err := getSystemGroupsInfoOfMonitoredClients(rpcClient, token, entitlement)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get the managed system groups information of monitored clients")
+	}
+
+	systemIDs := make([]int, 0, len(systemGroupIDsBySystemID))
+	for systemID := range systemGroupIDsBySystemID {
+		systemIDs = append(systemIDs, systemID)
+	}
+
+	endpointInfos, err := getEndpointInfoForSystems(rpcClient, token, systemIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get endpoints information")
+	}
+
+	networkInfoBySystemID, err := getNetworkInformationForSystems(rpcClient, token, systemIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get the systems network information")
+	}
+
+	for _, endpoint := range endpointInfos {
+		systemID := endpoint.SystemID
+		labels := d.getEndpointLabels(
+			endpoint,
+			systemGroupIDsBySystemID[systemID],
+			networkInfoBySystemID[systemID])
+		result = append(result, labels)
+	}
+
+	return result, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	cfg := d.sdConfig
+	apiURL := cfg.Host
+	apiURL.Path += uyuniXMLRPCAPIPath
```

Comment:
It seems like you are taking a pointer and at each refresh we add that to the origin SDConfig url :)

Reply:
Good catch. Thanks :)